diff -up openssh-9.1p1/openbsd-compat/openssl-compat.h.chachaBroken~ openssh-9.1p1/openbsd-compat/openssl-compat.h
--- openssh-9.1p1/openbsd-compat/openssl-compat.h.chachaBroken~	2022-12-24 14:25:14.781627554 +0100
+++ openssh-9.1p1/openbsd-compat/openssl-compat.h	2022-12-24 14:26:22.058836697 +0100
@@ -66,6 +66,8 @@ void ssh_libcrypto_init(void);
 # if LIBRESSL_VERSION_NUMBER < 0x3010000fL
 #  define HAVE_BROKEN_CHACHA20
 # endif
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000L
+# define HAVE_BROKEN_CHACHA20
 #endif
 
 /* LibreSSL/OpenSSL 1.1x API compat */
