diff -up openssh-9.0p1/ssh-keyscan.c.sigpipe openssh-9.0p1/ssh-keyscan.c
--- openssh-9.0p1/ssh-keyscan.c.sigpipe	2022-06-20 21:02:37.341127568 +0200
+++ openssh-9.0p1/ssh-keyscan.c	2022-06-20 21:06:46.686406567 +0200
@@ -776,6 +776,9 @@ main(int argc, char **argv)
 	if (maxfd > fdlim_get(0))
 		fdlim_set(maxfd);
 	fdcon = xcalloc(maxfd, sizeof(con));
+
+	signal(SIGPIPE, SIG_IGN);
+
 	read_wait = xcalloc(maxfd, sizeof(struct pollfd));
 	for (j = 0; j < maxfd; j++)
 		read_wait[j].fd = -1;
