diff -up openssh-9.3p1/sandbox-seccomp-filter.c.sandbox openssh-9.3p1/sandbox-seccomp-filter.c
--- openssh-9.3p1/sandbox-seccomp-filter.c.sandbox	2023-03-15 22:28:19.000000000 +0100
+++ openssh-9.3p1/sandbox-seccomp-filter.c	2023-05-30 20:49:31.599362791 +0200
@@ -277,6 +277,9 @@ static const struct sock_filter preauth_
 #ifdef __NR_exit_group
 	SC_ALLOW(__NR_exit_group),
 #endif
+#if defined(__NR_flock) && defined(__s390__)
+	SC_ALLOW(__NR_flock),
+#endif
 #ifdef __NR_futex
 	SC_FUTEX(__NR_futex),
 #endif
@@ -295,6 +298,21 @@ static const struct sock_filter preauth_
 #ifdef __NR_getpid
 	SC_ALLOW(__NR_getpid),
 #endif
+#ifdef __NR_getuid
+	SC_ALLOW(__NR_getuid),
+#endif
+#ifdef __NR_getuid32
+	SC_ALLOW(__NR_getuid32),
+#endif
+#ifdef __NR_geteuid
+	SC_ALLOW(__NR_geteuid),
+#endif
+#ifdef __NR_geteuid32
+	SC_ALLOW(__NR_geteuid32),
+#endif
+#ifdef __NR_gettid
+	SC_ALLOW(__NR_gettid),
+#endif
 #ifdef __NR_getrandom
 	SC_ALLOW(__NR_getrandom),
 #endif
@@ -304,6 +322,9 @@ static const struct sock_filter preauth_
 #ifdef __NR_gettimeofday
 	SC_ALLOW(__NR_gettimeofday),
 #endif
+#if defined(__NR_ipc) && defined(__s390__)
+	SC_ALLOW(__NR_ipc),
+#endif
 #ifdef __NR_getuid
 	SC_ALLOW(__NR_getuid),
 #endif
