diff -up openssh-8.4p1/openbsd-compat/openssl-compat.h.omv~ openssh-8.4p1/openbsd-compat/openssl-compat.h
--- openssh-8.4p1/openbsd-compat/openssl-compat.h.omv~	2020-10-03 01:29:45.010486770 +0200
+++ openssh-8.4p1/openbsd-compat/openssl-compat.h	2020-10-03 01:30:10.733901710 +0200
@@ -66,6 +66,8 @@ void ssh_libcrypto_init(void);
 # if LIBRESSL_VERSION_NUMBER < 0x3010000fL
 #  define HAVE_BROKEN_CHACHA20
 # endif
+#elif OPENSSL_VERSION_NUMBER >= 0x30000000L
+# define HAVE_BROKEN_CHACHA20
 #endif
 
 #ifndef OPENSSL_HAVE_EVPCTR
